swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt

```diff
@@ -0,0 +1,165 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.swabbie.notifications
+
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.kork.lock.LockManager
+import com.netflix.spinnaker.swabbie.LockingService
+import com.netflix.spinnaker.swabbie.discovery.DiscoveryAware
+import com.netflix.spinnaker.swabbie.events.OwnerNotifiedEvent
+import com.netflix.spinnaker.swabbie.events.UnMarkResourceEvent
+import com.netflix.spinnaker.swabbie.model.MarkedResource
+import com.netflix.spinnaker.swabbie.model.NotificationInfo
+import com.netflix.spinnaker.swabbie.model.WorkConfiguration
+import com.netflix.spinnaker.swabbie.repository.ResourceTrackingRepository
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Component
+import java.time.Clock
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+
+/**
+ * A scheduled notification sender
+ */
+@Component
+@ConditionalOnExpression("\${swabbie.enabled:true}")
+class NotificationSender(
+  private val lockingService: LockingService,
+  private val resourceTrackingRepository: ResourceTrackingRepository,
+  private val notifier: Notifier,
+  private val clock: Clock,
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val notificationQueue: NotificationQueue,
+  private val registry: Registry,
+  private val workConfigurations: List<WorkConfiguration>
+) : DiscoveryAware() {
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val notificationsId = registry.createId("swabbie.notifications")
+
+  private val lockOptions = LockManager.LockOptions()
+    .withLockName(lockingService.ownerName)
+    .withMaximumLockDuration(lockingService.swabbieMaxLockDuration)
+
+  /**
+   * Periodically sends notifications to resolved resource owners
+   * The frequency of this scheduled action is cron based and configurable
+   * Defaults to every day at 9am.
+   */
+  @Scheduled(cron = "\${swabbie.notification.cron.schedule:0 0 9 1/1 * ?}")
+  fun sendNotifications() {
+    if (!isUp()) {
+      return
+    }
+
+    lockingService.acquireLock(lockOptions) {
+      val resources = resourceTrackingRepository.getMarkedResources()
+        .filter {
+          it.notificationInfo == null
+        }
+
+      groupRelatedResourcesThenNotify(resources)
+    }
+  }
+
+  /**
+   * @return Marked resources to notify on
+   */
+  private fun groupRelatedResourcesThenNotify(markedResources: List<MarkedResource>) {
+    val now = Instant.now(clock)
+    val tasks = mutableListOf<NotificationTask>()
+    do {
+      val notificationTask: NotificationTask? = notificationQueue.pop()
+      if (notificationTask == null) {
```

Comment:
Could we just check `notificationQueue.isEmpty()` here?

Reply:
I guess, what's the case where the condition in this `do/while` loop would let this run but then there wouldn't be anything to notify? Are we locking on something else that would allow this to happen?

Reply:
A `do while` construct allows allows to check for that condition once. Otherwise you would have to check if the queue is empty, then proceed to pop items off of it and continue to check if the queue is empty.

Reply:
Not sure I follow the second part of the comment: `Are we locking on something else that would allow this to happen?`

Reply:
I was confused reading this at first because I didn't see the `do/while` and I didn't go back and update my comment. What if we just had a `popAll` method on the queue that returned everything?